source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -0,0 +1,138 @@
+#include "extensions/filters/http/cache/http_cache.h"
+
+#include <algorithm>
+
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/cache/http_cache_utils.h"
+
+#include "absl/time/time.h"
+
+using absl::Duration;
+using absl::FromChrono;
+using absl::ZeroDuration;
+using Envoy::Extensions::HttpFilters::Cache::Internal::effectiveMaxAge;
+using Envoy::Extensions::HttpFilters::Cache::Internal::httpTime;
+using Envoy::Http::HeaderEntry;
+using Envoy::Http::HeaderMap;
+using Envoy::Http::HeaderMapPtr;
+using Envoy::Http::HeaderString;
+using Envoy::Http::LowerCaseString;
+using std::string;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+LookupRequest::LookupRequest(const HeaderMap& request_headers, SystemTime timestamp)
+    : timestamp_(timestamp),
+      request_cache_control_(request_headers.CacheControl() == nullptr
+                                 ? ""
+                                 : request_headers.CacheControl()->value().getStringView()) {
+  RELEASE_ASSERT(request_headers.Path(),
+                 "Can't form cache lookup key for malformed HeaderMap with null Path.");
+  RELEASE_ASSERT(request_headers.Scheme(),
+                 "Can't form cache lookup key for malformed HeaderMap with null Scheme.");
+  RELEASE_ASSERT(request_headers.Host(),
+                 "Can't form cache lookup key for malformed HeaderMap with null Host.");
+  // TODO(toddmgreer) Let config determine whether to include scheme, host, and
+  // query params.
+  // TODO(toddmgreer) get cluster name.
+  // TODO(toddmgreer) Parse Range header into request_range_spec_.
+  key_.set_cluster_name("cluster_name_goes_here");
+  key_.set_host(string(request_headers.Host()->value().getStringView()));
+  key_.set_path(string(request_headers.Path()->value().getStringView()));
+  const HeaderString& scheme = request_headers.Scheme()->value();
+  ASSERT(scheme == "http" || scheme == "https");
+  key_.set_clear_http(scheme == "http");
+}
+
+// Unless this API is still alpha, calls to stableHashKey() must always return
+// the same result, or a way must be provided to deal with a complete cache
+// flush. localHashKey however, can be changed at will.
+size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
+size_t localHashKey(const Key& key) { return stableHashKey(key); }
+
+// Returns true if response_headers is fresh.
+bool LookupRequest::fresh(const HeaderMap& response_headers) const {
+  const HeaderEntry* cache_control_header = response_headers.get(LowerCaseString("cache-control"));
```

Comment:
```
  const HeaderEntry* cache_control_header = resposne_headers.CacheControl();
```
There's macro magic to directly and quickly exact header values for many common header names.

Reply:
Done.